mono/metadata/mempool.c

```diff
@@ -254,6 +274,102 @@ get_next_size (MonoMemPool *pool, int size)
 #endif
 
 /**
+ * mono_mempool_enable_canaries:
+ *
+ * Enable placing canaries between allocations
+ */
+
+void
+mono_mempool_enable_canaries (void)
+{
+	canary_count = DEFAULT_ENABLED_CANARY_COUNT;
+	return;
+}
+
+/**
+ * mono_mempool_check_canaries:
+ *
+ * Check that our canaries are still alive.
+ */
+
+void
+mono_mempool_check_canaries (MonoMemPool *pool)
+{
+	if (canary_count == 0)
+		return;
+
+#ifdef MALLOC_ALLOCATION
+	for (Chunk* c = pool->chunks; c != NULL; c++) {
+		gpointer start = c + sizeof(Chunk) + c->size;
+		gpointer end = start + PADDING_COUNT;
+
+		// The end should be the value stored at the start of the canary
+		assert(end == start + PADDING_COUNT + 1);
+
+		for (gpointer addr = start + sizeof(gpointer); addr < end; addr++) {
+		    assert(*addr == CANARY_VALUE);
+		}
+	}
+#else
+	MonoMemPool *curr_pool;
+	AllocRecord *curr;
+
+	for (curr_pool = pool; curr_pool != NULL; curr_pool = curr_pool->next) {
+		for (curr = curr_pool->alloc_records; curr != NULL; curr = curr->next) {
+			for (int i=0; i < canary_count; i++){
+#if 0
+// Insert earlier for sleep 
+// #include <unistd.h>
```

Comment:
Instead of #if 0, give it a name and add on the top of the file a list of knobs to enable and what they do.